[clang-tidy] add modernize-use-constexpr check (#146553)
llvm/llvm-project: clang-tools-extra/docs/clang-tidy/checks/modernize/use-constexpr.rst

.. title:: clang-tidy - modernize-use-constexpr

modernize-use-constexpr
=======================

> Finds functions and variables that can be declared 'constexpr'.

Reviewer: ```suggestion
Finds functions and variables that can be declared '`constexpr'`.
```

This is because this check would have to duplicate the expensive analysis of the
:doc:`misc-const-correctness<../misc/const-correctness>` check.
Therefore, it is recommended to have 
:doc:`misc-const-correctness<../misc/const-correctness>` enabled
in the Clang-Tidy config when this check is, so that all opportunities for
``const`` and also ``constexpr`` are explored.

Options
-------

.. option:: ConservativeLiteralType

  With this option enabled, only literal types that can be constructed at
  compile-time are considered to supoprt ``constexpr``.

  .. code-block:: c++

    struct NonLiteral{
      NonLiteral();
      ~NonLiteral();
      int &ref;
    };

  This type is a literal type, but can not be constructed at compile-time.
  With `ConservativeLiteralType` equal to `true`, variables or funtions
  with this type are not diagnosed to add ``constexpr``. Default is
  `true`.

.. option:: AddConstexprToMethodOfClassWithoutConstexprConstructor

  While a function of a class or struct could be declared ``constexpr``, when
  the class itself can never be constructed at compile-time, then adding
  ``constexpr`` to a member function is superfluous. This option controls if

>   ``constexpr`` should be added anyways. Default is ``false``.

Reviewer: ```suggestion
  ``constexpr`` should be added anyways. Default is `false`.
```

.. option:: ConstexprString

  The string to use to specify a variable or function as ``constexpr``, for

>   example, a macro. Default is ``constexpr``.

Reviewer: ```suggestion
  example, a macro. Default is `constexpr`.
```

.. option:: ConstexprString

  The string to use with C++23 to specify a function-local variable as 

>   ``static constexpr``, for example, a macro. Default is ``static constexpr``

Reviewer: ```suggestion
  ``static constexpr``, for example, a macro. Default is `static constexpr`
```
